diff -u -r data/changesInfo//Codec_3/properties/modified_classes/inducing/src/java/org/apache/commons/codec/language/DoubleMetaphone.java data/changesInfo//Codec_3/cleaned/fixing/src/java/org/apache/commons/codec/language/DoubleMetaphone.java
--- data/changesInfo//Codec_3/properties/modified_classes/inducing/src/java/org/apache/commons/codec/language/DoubleMetaphone.java	2023-07-26 02:09:14.951645272 -0700
+++ data/changesInfo//Codec_3/cleaned/fixing/src/java/org/apache/commons/codec/language/DoubleMetaphone.java	2023-07-26 02:18:21.643026316 -0700
@@ -451,7 +451,7 @@
                 contains(value, index + 1, 2, "ET")) {
                 //-- obvious germanic --//
                 result.append('K');
-            } else if (contains(value, index + 1, 4, "IER")) {
+            } else if (contains(value, index + 1, 2, "IER")) {
                 result.append('J');
             } else {
                 result.append('J', 'K');
@@ -560,14 +560,16 @@
      * Handles 'L' cases
      */
     private int handleL(String value, DoubleMetaphoneResult result, int index) {
-        result.append('L');
         if (charAt(value, index + 1) == 'L') {
             if (conditionL0(value, index)) {
-                result.appendAlternate(' ');
+                result.appendPrimary('L');
+            } else {
+                result.append('L');
             }
             index += 2;
         } else {
             index++;
+            result.append('L');
         }
         return index;
     }
@@ -855,7 +857,7 @@
         if (index == value.length() - 3 && 
             contains(value, index - 1, 4, "ILLO", "ILLA", "ALLE")) {
             return true;
-        } else if ((contains(value, index - 1, 2, "AS", "OS") || 
+        } else if ((contains(value, value.length() - 2, 2, "AS", "OS") ||
                     contains(value, value.length() - 1, 1, "A", "O")) &&
                    contains(value, index - 1, 4, "ALLE")) {
             return true;
Only in data/changesInfo//Codec_3/properties/modified_classes/inducing/src: test
