diff -r -u buggy/src/com/google/javascript/jscomp/CheckGlobalThis.java inducing/src/com/google/javascript/jscomp/CheckGlobalThis.java
--- buggy/src/com/google/javascript/jscomp/CheckGlobalThis.java	2023-06-12 11:34:30.706298312 +0800
+++ inducing/src/com/google/javascript/jscomp/CheckGlobalThis.java	2023-06-12 11:34:31.042304451 +0800
@@ -53,6 +53,8 @@
  * </pre>
  * which would get flagged.
  *
+*
+*
  */
 final class CheckGlobalThis implements Callback {
 
@@ -104,10 +106,7 @@
             pType == Token.SCRIPT ||
             pType == Token.NAME ||
             pType == Token.ASSIGN ||
-
-            // object literal keys
-            pType == Token.STRING ||
-            pType == Token.NUMBER)) {
+            pType == Token.OBJECTLIT)) {
         return false;
       }
     }
