diff -r -u inducing/src/com/google/javascript/jscomp/CheckGlobalThis.java buggy/src/com/google/javascript/jscomp/CheckGlobalThis.java
--- inducing/src/com/google/javascript/jscomp/CheckGlobalThis.java	2023-06-12 11:40:28.410625475 +0800
+++ buggy/src/com/google/javascript/jscomp/CheckGlobalThis.java	2023-06-12 11:40:28.234633548 +0800
@@ -53,8 +53,6 @@
  * </pre>
  * which would get flagged.
  *
-*
-*
  */
 final class CheckGlobalThis implements Callback {
 
@@ -106,7 +104,10 @@
             pType == Token.SCRIPT ||
             pType == Token.NAME ||
             pType == Token.ASSIGN ||
-            pType == Token.OBJECTLIT)) {
+
+            // object literal keys
+            pType == Token.STRING ||
+            pType == Token.NUMBER)) {
         return false;
       }
     }
