diff --git a/src/java/org/apache/commons/lang/LocaleUtils.java b/src/java/org/apache/commons/lang/LocaleUtils.java
index 139d9a7..9607953 100644
--- a/src/java/org/apache/commons/lang/LocaleUtils.java
+++ b/src/java/org/apache/commons/lang/LocaleUtils.java
@@ -111,6 +111,9 @@
                 throw new IllegalArgumentException("Invalid locale format: " + str);
             }
             char ch3 = str.charAt(3);
+            if (ch3 == '_') {
+                return new Locale(str.substring(0, 2), "", str.substring(4));
+            }
             char ch4 = str.charAt(4);
             if (ch3 < 'A' || ch3 > 'Z' || ch4 < 'A' || ch4 > 'Z') {
                 throw new IllegalArgumentException("Invalid locale format: " + str);
diff --git a/src/test/org/apache/commons/lang/LocaleUtilsTest.java b/src/test/org/apache/commons/lang/LocaleUtilsTest.java
index 1855544..2c4324c 100644
--- a/src/test/org/apache/commons/lang/LocaleUtilsTest.java
+++ b/src/test/org/apache/commons/lang/LocaleUtilsTest.java
@@ -498,4 +498,11 @@
         } catch (UnsupportedOperationException ex) {}
     }
 
+    /**
+     * Tests #LANG-328 - only language+variant
+     */
+    public void testLang328() {
+        assertValidToLocale("fr__POSIX", "fr", "", "POSIX");
+    }
+
 }
