diff -u -r data/changesInfo//Closure_62/properties/modified_classes/inducing/src/com/google/javascript/jscomp/LightweightMessageFormatter.java data/changesInfo//Closure_62/cleaned/fixing/src/com/google/javascript/jscomp/LightweightMessageFormatter.java
--- data/changesInfo//Closure_62/properties/modified_classes/inducing/src/com/google/javascript/jscomp/LightweightMessageFormatter.java	2023-12-27 22:39:14.681812871 -0800
+++ data/changesInfo//Closure_62/cleaned/fixing/src/com/google/javascript/jscomp/LightweightMessageFormatter.java	2023-06-20 05:30:08.973168501 -0700
@@ -93,7 +93,7 @@
 
       // padding equal to the excerpt and arrow at the end
       if (excerpt.equals(LINE)
-          && 0 <= charno && charno < sourceExcerpt.length()) {
+          && 0 <= charno && charno <= sourceExcerpt.length()) {
         for (int i = 0; i < charno; i++) {
           char c = sourceExcerpt.charAt(i);
           if (Character.isWhitespace(c)) {
Only in data/changesInfo//Closure_62/properties/modified_classes/inducing/: test
