diff -u -r data/changesInfo//Closure_107/properties/modified_classes/inducing/src/com/google/javascript/jscomp/CommandLineRunner.java data/changesInfo//Closure_107/cleaned/fixing/src/com/google/javascript/jscomp/CommandLineRunner.java
--- data/changesInfo//Closure_107/properties/modified_classes/inducing/src/com/google/javascript/jscomp/CommandLineRunner.java	2023-12-27 22:43:04.537209676 -0800
+++ data/changesInfo//Closure_107/cleaned/fixing/src/com/google/javascript/jscomp/CommandLineRunner.java	2023-07-25 00:26:14.681019022 -0700
@@ -666,6 +666,7 @@
       // In ADVANCED mode, goog.getMsg is going to be renamed anyway,
       // so we might as well inline it.
       options.messageBundle = new EmptyMessageBundle();
+      options.setWarningLevel(JsMessageVisitor.MSG_CONVENTIONS, CheckLevel.OFF);
     }
 
     return options;
diff -u -r data/changesInfo//Closure_107/properties/modified_classes/inducing/src/com/google/javascript/jscomp/JsMessageVisitor.java data/changesInfo//Closure_107/cleaned/fixing/src/com/google/javascript/jscomp/JsMessageVisitor.java
--- data/changesInfo//Closure_107/properties/modified_classes/inducing/src/com/google/javascript/jscomp/JsMessageVisitor.java	2023-07-25 00:42:16.088454884 -0700
+++ data/changesInfo//Closure_107/cleaned/fixing/src/com/google/javascript/jscomp/JsMessageVisitor.java	2023-07-25 00:43:56.810894694 -0700
@@ -71,6 +71,20 @@
       DiagnosticType.error("JSC_MSG_NOT_INITIALIZED_USING_NEW_SYNTAX",
           "message not initialized using " + MSG_FUNCTION_NAME);
 
+  /**
+   * Warnings that only apply to people who use MSG_ to denote
+   * messages. Note that this doesn't include warnings about
+   * proper use of goog.getMsg
+   */
+  static final DiagnosticGroup MSG_CONVENTIONS = new DiagnosticGroup(
+      "messageConventions",
+      MESSAGE_HAS_NO_DESCRIPTION,
+      MESSAGE_HAS_NO_TEXT,
+      MESSAGE_TREE_MALFORMED,
+      MESSAGE_HAS_NO_VALUE,
+      MESSAGE_DUPLICATE_KEY,
+      MESSAGE_NOT_INITIALIZED_USING_NEW_SYNTAX);
+
   private static final String PH_JS_PREFIX = "{$";
   private static final String PH_JS_SUFFIX = "}";
 
Only in data/changesInfo//Closure_107/properties/modified_classes/inducing/: test
