diff -u -r data/changesInfo//Closure_125/properties/modified_classes/inducing/src/com/google/javascript/jscomp/TypeCheck.java data/changesInfo//Closure_125/cleaned/fixing/src/com/google/javascript/jscomp/TypeCheck.java
--- data/changesInfo//Closure_125/properties/modified_classes/inducing/src/com/google/javascript/jscomp/TypeCheck.java	2023-07-25 04:05:36.599519943 -0700
+++ data/changesInfo//Closure_125/cleaned/fixing/src/com/google/javascript/jscomp/TypeCheck.java	2023-07-27 05:34:52.297021578 -0700
@@ -1284,9 +1284,11 @@
   private void visitNew(NodeTraversal t, Node n) {
     Node constructor = n.getFirstChild();
     FunctionType type = getFunctionType(constructor);
-    if (type != null && type.isConstructor()) {
-      visitParameterList(t, n, type);
-      ensureTyped(t, n, type.getInstanceType());
+    if (type != null && (type.isConstructor() || type.isEmptyType() || type.isUnknownType())) {
+      if (type.hasInstanceType()) {
+        visitParameterList(t, n, type);
+        ensureTyped(t, n, type.getInstanceType());
+      }
     } else {
       // TODO(user): add support for namespaced objects.
       if (constructor.getType() != Token.GETPROP) {
@@ -1301,8 +1303,8 @@
         }
         report(t, line, NOT_A_CONSTRUCTOR);
       }
-      ensureTyped(t, n);
     }
+    ensureTyped(t, n);
   }
 
   /**
Only in data/changesInfo//Closure_125/properties/modified_classes/inducing/src/com/google/javascript: rhino
Only in data/changesInfo//Closure_125/properties/modified_classes/inducing/: test
