diff -r -u inducing/src/main/java/org/apache/commons/lang3/SerializationUtils.java buggy/src/main/java/org/apache/commons/lang3/SerializationUtils.java
--- inducing/src/main/java/org/apache/commons/lang3/SerializationUtils.java	2023-06-12 11:40:35.826288031 +0800
+++ buggy/src/main/java/org/apache/commons/lang3/SerializationUtils.java	2023-06-12 11:40:35.718292907 +0800
@@ -89,7 +89,9 @@
              * it is reasonable to assume the deserialized object
              * is of the same type as the original serialized object
              */
-            return (T) in.readObject();
+            @SuppressWarnings("unchecked") // see above
+            T readObject = (T) in.readObject();
+            return readObject;
 
         } catch (ClassNotFoundException ex) {
             throw new SerializationException("ClassNotFoundException while reading cloned object data", ex);
diff -r -u inducing/src/test/java/org/apache/commons/lang3/SerializationUtilsTest.java buggy/src/test/java/org/apache/commons/lang3/SerializationUtilsTest.java
--- inducing/src/test/java/org/apache/commons/lang3/SerializationUtilsTest.java	2023-06-12 11:40:35.826288031 +0800
+++ buggy/src/test/java/org/apache/commons/lang3/SerializationUtilsTest.java	2023-06-12 11:40:35.738292004 +0800
@@ -53,7 +53,7 @@
         super.setUp();
 
         iString = "foo";
-        iInteger = new Integer(7);
+        iInteger = Integer.valueOf(7);
         iMap = new HashMap<Object, Object>();
         iMap.put("FOO", iString);
         iMap.put("BAR", iInteger);
