diff -u -r data/changesInfo//Mockito_11/properties/modified_classes/inducing/src/org/mockito/internal/creation/DelegatingMethod.java data/changesInfo//Mockito_11/cleaned/fixing/src/org/mockito/internal/creation/DelegatingMethod.java
--- data/changesInfo//Mockito_11/properties/modified_classes/inducing/src/org/mockito/internal/creation/DelegatingMethod.java	2023-07-24 01:12:09.331269500 -0700
+++ data/changesInfo//Mockito_11/cleaned/fixing/src/org/mockito/internal/creation/DelegatingMethod.java	2023-07-24 01:53:57.935374013 -0700
@@ -44,6 +44,13 @@
     
     @Override
     public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj instanceof DelegatingMethod) {
+            DelegatingMethod that = (DelegatingMethod) obj;
+            return method.equals(that.method);
+        }
         return method.equals(obj);
     }
 }
\ No newline at end of file
Only in data/changesInfo//Mockito_11/properties/modified_classes/inducing/src/org/mockito/internal/creation: MethodInterceptorFilter.java
Only in data/changesInfo//Mockito_11/properties/modified_classes/inducing/src/org/mockito/internal: invocation
Only in data/changesInfo//Mockito_11/properties/modified_classes/inducing/src/org/mockito/internal: util
Only in data/changesInfo//Mockito_11/properties/modified_classes/inducing/: test
