diff -u -r data/changesInfo//Closure_115/properties/modified_classes/inducing/src/com/google/javascript/jscomp/FunctionInjector.java data/changesInfo//Closure_115/cleaned/fixing/src/com/google/javascript/jscomp/FunctionInjector.java
--- data/changesInfo//Closure_115/properties/modified_classes/inducing/src/com/google/javascript/jscomp/FunctionInjector.java	2023-12-27 22:43:49.061867393 -0800
+++ data/changesInfo//Closure_115/cleaned/fixing/src/com/google/javascript/jscomp/FunctionInjector.java	2023-06-15 03:28:42.009285394 -0700
@@ -694,16 +694,6 @@
 
     Node block = fnNode.getLastChild();
 
-    boolean hasSideEffects = false;  // empty function case
-    if (block.hasChildren()) {
-      Preconditions.checkState(block.hasOneChild());
-      Node stmt = block.getFirstChild();
-      if (stmt.isReturn()) {
-        hasSideEffects = NodeUtil.mayHaveSideEffects(
-            stmt.getFirstChild(), compiler);
-      }
-    }
-
     // CALL NODE: [ NAME, ARG1, ARG2, ... ]
     Node cArg = callNode.getFirstChild().getNext();
 
@@ -729,10 +719,6 @@
       // For each named parameter check if a mutable argument use more than one.
       if (fnParam != null) {
         if (cArg != null) {
-          if (hasSideEffects && NodeUtil.canBeSideEffected(cArg)) {
-            return CanInlineResult.NO;
-          }
-
           // Check for arguments that are evaluated more than once.
           // Note: Unlike block inlining, there it is not possible that a
           // parameter reference will be in a loop.
Only in data/changesInfo//Closure_115/properties/modified_classes/inducing/: test
