diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index b58d95f..0216cd0 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -23,6 +23,12 @@
   </properties>
   <body>
     <release version="1.1" date="as in SVN" description="Release 1.1">
+      <action type="fix" issue="COMPRESS-94" date="2010-01-07"
+              due-to="Anon Devs">
+        ZipArchiveEntry's equals method was broken for entries created
+        with the String-arg constructor.  This lead to broken ZIP
+        archives if two different entries had the same hash code.
+      </action>
       <action type="fix" issue="COMPRESS-87" date="2009-10-30"
               due-to="Antoni Mylka">
         ZipArchiveInputStream could repeatedly return 0 on read() when
diff --git a/src/main/java/org/apache/commons/compress/archivers/zip/ZipArchiveEntry.java b/src/main/java/org/apache/commons/compress/archivers/zip/ZipArchiveEntry.java
index cab4c1c..fe5270f 100644
--- a/src/main/java/org/apache/commons/compress/archivers/zip/ZipArchiveEntry.java
+++ b/src/main/java/org/apache/commons/compress/archivers/zip/ZipArchiveEntry.java
@@ -61,6 +61,7 @@
      */
     public ZipArchiveEntry(String name) {
         super(name);
+        setName(name);
     }
 
     /**
@@ -96,7 +97,7 @@
     /**
      */
     protected ZipArchiveEntry() {
-        super("");
+        this("");
     }
 
     public ZipArchiveEntry(File inputFile, String entryName) {
@@ -459,11 +460,13 @@
             return false;
         }
         ZipArchiveEntry other = (ZipArchiveEntry) obj;
-        if (name == null) {
-            if (other.name != null) {
+        String myName = getName();
+        String otherName = other.getName();
+        if (myName == null) {
+            if (otherName != null) {
                 return false;
             }
-        } else if (!name.equals(other.name)) {
+        } else if (!myName.equals(otherName)) {
             return false;
         }
         return true;
diff --git a/src/test/java/org/apache/commons/compress/archivers/zip/ZipArchiveEntryTest.java b/src/test/java/org/apache/commons/compress/archivers/zip/ZipArchiveEntryTest.java
index c50defe..1ce7822 100644
--- a/src/test/java/org/apache/commons/compress/archivers/zip/ZipArchiveEntryTest.java
+++ b/src/test/java/org/apache/commons/compress/archivers/zip/ZipArchiveEntryTest.java
@@ -219,4 +219,14 @@
         assertFalse(entry.isSupportedCompressionMethod());
     }
 
+    /**
+     * Test case for
+     * <a href="https://issues.apache.org/jira/browse/COMPRESS-94"
+     * >COMPRESS-94</a>.
+     */
+    public void testNotEquals() {
+        ZipArchiveEntry entry1 = new ZipArchiveEntry("foo");
+        ZipArchiveEntry entry2 = new ZipArchiveEntry("bar");
+        assertFalse(entry1.equals(entry2));
+    }
 }
