diff -r -u buggy/src/test/java/org/apache/commons/math/analysis/solvers/RegulaFalsiSolverTest.java inducing/src/test/java/org/apache/commons/math/analysis/solvers/RegulaFalsiSolverTest.java
--- buggy/src/test/java/org/apache/commons/math/analysis/solvers/RegulaFalsiSolverTest.java	2023-06-12 11:34:48.774628184 +0800
+++ inducing/src/test/java/org/apache/commons/math/analysis/solvers/RegulaFalsiSolverTest.java	2023-06-12 11:34:49.350638693 +0800
@@ -43,7 +43,7 @@
     @Test
     public void testIssue631() {
         final UnivariateRealFunction f = new UnivariateRealFunction() {
-                /** {@inheritDoc} */
+                @Override
                 public double value(double x) {
                     return Math.exp(x) - Math.pow(Math.PI, 3.0);
                 }
