diff -r -u inducing/src/test/java/org/apache/commons/math/analysis/solvers/RegulaFalsiSolverTest.java buggy/src/test/java/org/apache/commons/math/analysis/solvers/RegulaFalsiSolverTest.java
--- inducing/src/test/java/org/apache/commons/math/analysis/solvers/RegulaFalsiSolverTest.java	2023-06-12 11:40:45.873839751 +0800
+++ buggy/src/test/java/org/apache/commons/math/analysis/solvers/RegulaFalsiSolverTest.java	2023-06-12 11:40:45.833841516 +0800
@@ -43,7 +43,7 @@
     @Test
     public void testIssue631() {
         final UnivariateRealFunction f = new UnivariateRealFunction() {
-                @Override
+                /** {@inheritDoc} */
                 public double value(double x) {
                     return Math.exp(x) - Math.pow(Math.PI, 3.0);
                 }
