Only in original/src: main
diff -r -u original/src/test/java/org/apache/commons/math/analysis/solvers/RegulaFalsiSolverTest.java buggy/src/test/java/org/apache/commons/math/analysis/solvers/RegulaFalsiSolverTest.java
--- original/src/test/java/org/apache/commons/math/analysis/solvers/RegulaFalsiSolverTest.java	2023-06-13 15:31:55.207159286 +0800
+++ buggy/src/test/java/org/apache/commons/math/analysis/solvers/RegulaFalsiSolverTest.java	2023-06-12 11:40:45.833841516 +0800
@@ -17,6 +17,10 @@
 
 package org.apache.commons.math.analysis.solvers;
 
+import org.apache.commons.math.analysis.UnivariateRealFunction;
+import org.junit.Test;
+import org.junit.Assert;
+
 /**
  * Test case for {@link RegulaFalsiSolver Regula Falsi} solver.
  *
@@ -35,4 +39,18 @@
         // even a million iterations. As such, it was disabled.
         return new int[] {3, 7, 8, 19, 18, 11, 67, 55, 288, 151, -1};
     }
+
+    @Test
+    public void testIssue631() {
+        final UnivariateRealFunction f = new UnivariateRealFunction() {
+                /** {@inheritDoc} */
+                public double value(double x) {
+                    return Math.exp(x) - Math.pow(Math.PI, 3.0);
+                }
+            };
+
+        final UnivariateRealSolver solver = new RegulaFalsiSolver();
+        final double root = solver.solve(3624, f, 1, 10);
+        Assert.assertEquals(3.4341896575482003, root, 1e-15);
+	}
 }
