diff -u -r data/changesInfo//Closure_113/properties/modified_classes/inducing/src/com/google/javascript/jscomp/ProcessClosurePrimitives.java data/changesInfo//Closure_113/cleaned/fixing/src/com/google/javascript/jscomp/ProcessClosurePrimitives.java
--- data/changesInfo//Closure_113/properties/modified_classes/inducing/src/com/google/javascript/jscomp/ProcessClosurePrimitives.java	2023-12-28 06:30:19.264304854 -0800
+++ data/changesInfo//Closure_113/cleaned/fixing/src/com/google/javascript/jscomp/ProcessClosurePrimitives.java	2023-06-15 01:35:27.527296820 -0700
@@ -322,11 +322,12 @@
       maybeAddStringNodeToSymbolTable(arg);
 
       // Requires should be removed before runtime.  The one
-      // exception is if the type has not been provided yet, in which case,
+      // exception is if the type has not been provided yet and
+      // errors for broken requires are turned off, in which case,
       // we will be doing a later pass that may error, so we can
       // leave this here this time and let it error next time if it
       // is still not provided.
-      if (provided != null) {
+      if (provided != null || requiresLevel.isOn()) {
         parent.detachFromParent();
         compiler.reportCodeChange();
       }
