diff --git a/doap_lang.rdf b/doap_lang.rdf
index 8662648..d4bf312 100644
--- a/doap_lang.rdf
+++ b/doap_lang.rdf
@@ -8,7 +8,7 @@
     <license rdf:resource="http://usefulinc.com/doap/licenses/asl20"/>
     <bug-database rdf:resource="http://issues.apache.org/bugzilla/buglist.cgi?product=Commons&amp;component=lang"/>
     <download-page rdf:resource="http://jakarta.apache.org/site/downloads/downloads_commons-lang.cgi"/>
-    <asfext:pmc rdf:resource="Jakarta"/>
+    <asfext:pmc rdf:resource="http://jakarta.apache.org/"/>
     <shortdesc xml:lang="en">Java Common Components</shortdesc>
     <description xml:lang="en">
         Commons.Lang, a package of Java utility classes for the
