diff -u -r data/changesInfo//Time_7/properties/modified_classes/inducing/JodaTime/src/main/java/org/joda/time/format/DateTimeFormatter.java data/changesInfo//Time_7/cleaned/fixing/JodaTime/src/main/java/org/joda/time/format/DateTimeFormatter.java
--- data/changesInfo//Time_7/properties/modified_classes/inducing/JodaTime/src/main/java/org/joda/time/format/DateTimeFormatter.java	2023-12-27 22:48:46.895277097 -0800
+++ data/changesInfo//Time_7/cleaned/fixing/JodaTime/src/main/java/org/joda/time/format/DateTimeFormatter.java	2023-06-24 03:32:04.051243104 -0700
@@ -699,9 +699,10 @@
         Chronology chrono = instant.getChronology();
         long instantLocal = instantMillis + chrono.getZone().getOffset(instantMillis);
         chrono = selectChronology(chrono);
+        int defaultYear = DateTimeUtils.getChronology(chrono).year().get(instantMillis);
         
         DateTimeParserBucket bucket = new DateTimeParserBucket(
-            instantLocal, chrono, iLocale, iPivotYear, iDefaultYear);
+            instantLocal, chrono, iLocale, iPivotYear, defaultYear);
         int newPos = parser.parseInto(bucket, text, position);
         instant.setMillis(bucket.computeMillis(false, text));
         if (iOffsetParsed && bucket.getZone() == null) {
Only in data/changesInfo//Time_7/properties/modified_classes/inducing/JodaTime/src/main/java/org/joda/time/format: DateTimeParserBucket.java
Only in data/changesInfo//Time_7/properties/modified_classes/inducing/JodaTime/src: test
Only in data/changesInfo//Time_7/properties/modified_classes/inducing/: src
