Only in buggy/src: main
diff -r -u inducing/src/test/java/org/apache/commons/lang3/text/translate/LookupTranslatorTest.java buggy/src/test/java/org/apache/commons/lang3/text/translate/LookupTranslatorTest.java
--- inducing/src/test/java/org/apache/commons/lang3/text/translate/LookupTranslatorTest.java	2023-06-12 11:40:35.262313498 +0800
+++ buggy/src/test/java/org/apache/commons/lang3/text/translate/LookupTranslatorTest.java	2023-06-12 11:40:35.190316750 +0800
@@ -15,23 +15,26 @@
  * limitations under the License.
  */
 
-package org.apache.commons.lang.text.translate;
+package org.apache.commons.lang3.text.translate;
 
-import junit.framework.TestCase;
+import static org.junit.Assert.assertEquals;
 
-import java.io.StringWriter;
 import java.io.IOException;
+import java.io.StringWriter;
+
+import org.junit.Test;
 
 /**
- * Unit tests for {@link org.apache.commons.lang.text.translate.LookupTranslator}.
+ * Unit tests for {@link org.apache.commons.lang3.text.translate.LookupTranslator}.
+ * @version $Id$
  */
-public class LookupTranslatorTest extends TestCase {
+public class LookupTranslatorTest  {
 
+    @Test
     public void testBasicLookup() throws IOException {
-        String input = "one";
-        LookupTranslator lt = new LookupTranslator(new CharSequence[][] { { "one", "two" } });
-        StringWriter out = new StringWriter();
-        int result = lt.translate("one", 0, out);
+        final LookupTranslator lt = new LookupTranslator(new CharSequence[][] { { "one", "two" } });
+        final StringWriter out = new StringWriter();
+        final int result = lt.translate("one", 0, out);
         assertEquals("Incorrect codepoint consumption", 3, result);
         assertEquals("Incorrect value", "two", out.toString());
     }
