diff -u -r data/changesInfo//Closure_114/properties/modified_classes/inducing/src/com/google/javascript/jscomp/NameAnalyzer.java data/changesInfo//Closure_114/cleaned/fixing/src/com/google/javascript/jscomp/NameAnalyzer.java
--- data/changesInfo//Closure_114/properties/modified_classes/inducing/src/com/google/javascript/jscomp/NameAnalyzer.java	2023-12-27 22:43:35.693669926 -0800
+++ data/changesInfo//Closure_114/cleaned/fixing/src/com/google/javascript/jscomp/NameAnalyzer.java	2023-07-25 01:00:28.767129908 -0700
@@ -514,7 +514,7 @@
             // logic.
             if (parent.getFirstChild().getNext() != n) {
               recordDepScope(n, ns);
-            } else {
+            } else if (!(parent.getType() == Token.CALL && parent.getFirstChild() == n)) {
               recordDepScope(nameNode, ns);
             }
           } else {
Only in data/changesInfo//Closure_114/properties/modified_classes/inducing/src/com/google/javascript: rhino
Only in data/changesInfo//Closure_114/properties/modified_classes/inducing/: test
