diff -u -r data/changesInfo//Closure_65/properties/modified_classes/inducing/test/com/google/javascript/jscomp/CodePrinterTest.java data/changesInfo//Closure_65/cleaned/inducing/test/com/google/javascript/jscomp/CodePrinterTest.java
--- data/changesInfo//Closure_65/properties/modified_classes/inducing/test/com/google/javascript/jscomp/CodePrinterTest.java	2023-06-11 20:40:24.302814833 -0700
+++ data/changesInfo//Closure_65/cleaned/inducing/test/com/google/javascript/jscomp/CodePrinterTest.java	2023-06-14 01:43:22.159552094 -0700
@@ -616,4 +616,12 @@
     assertPrint("if(e1)function goo(){return true}",
         "if(e1){function goo(){return true}}");
   }
+
+  public void testZero() {
+    assertPrint("var x ='\\0';", "var x=\"\\000\"");
+    assertPrint("var x ='\\x00';", "var x=\"\\000\"");
+    assertPrint("var x ='\\u0000';", "var x=\"\\000\"");
+    assertPrint("var x ='\\u00003';", "var x=\"\\0003\"");
+  }
+
 }
