diff -u -r data/changesInfo//Math_15/properties/modified_classes/inducing/src/main/java/org/apache/commons/math/util/FastMath.java data/changesInfo//Math_15/cleaned/fixing/src/main/java/org/apache/commons/math/util/FastMath.java
--- data/changesInfo//Math_15/properties/modified_classes/inducing/src/main/java/org/apache/commons/math/util/FastMath.java	2023-12-27 22:46:28.732225501 -0800
+++ data/changesInfo//Math_15/cleaned/fixing/src/main/java/org/apache/commons/math/util/FastMath.java	2023-06-20 07:19:40.444978476 -0700
@@ -1323,7 +1323,7 @@
         /* Handle special case x<0 */
         if (x < 0) {
             // y is an even integer in this case
-            if (y >= 4503599627370496.0 || y <= -4503599627370496.0) {
+            if (y >= 2 * 4503599627370496.0 || y <= -(2 * 4503599627370496.0)) {
                 return pow(-x, y);
             }
 
