diff -u -r data/changesInfo//Closure_128/properties/modified_classes/inducing/src/com/google/javascript/jscomp/CodeGenerator.java data/changesInfo//Closure_128/cleaned/fixing/src/com/google/javascript/jscomp/CodeGenerator.java
--- data/changesInfo//Closure_128/properties/modified_classes/inducing/src/com/google/javascript/jscomp/CodeGenerator.java	2023-12-27 22:44:47.434729570 -0800
+++ data/changesInfo//Closure_128/cleaned/fixing/src/com/google/javascript/jscomp/CodeGenerator.java	2023-06-20 06:38:50.182621631 -0700
@@ -742,7 +742,7 @@
         return false;
       }
     }
-    return len > 0 && s.charAt(0) != '0';
+    return len > 0 && (len == 1 || s.charAt(0) != '0');
   }
 
   static double getSimpleNumber(String s) {
Only in data/changesInfo//Closure_128/properties/modified_classes/inducing/: test
