diff -r -u buggy/src/com/google/javascript/jscomp/Compiler.java inducing/src/com/google/javascript/jscomp/Compiler.java
--- buggy/src/com/google/javascript/jscomp/Compiler.java	2023-06-12 11:34:23.062158608 +0800
+++ inducing/src/com/google/javascript/jscomp/Compiler.java	2023-06-12 11:34:23.270162410 +0800
@@ -24,9 +24,6 @@
 import com.google.javascript.jscomp.CompilerOptions.DevMode;
 import com.google.javascript.jscomp.CompilerOptions.LanguageMode;
 import com.google.javascript.jscomp.CompilerOptions.TracerMode;
-import com.google.javascript.jscomp.ReferenceCollectingCallback.ReferenceCollection;
-import com.google.javascript.jscomp.ReferenceCollectingCallback.ReferenceMap;
-import com.google.javascript.jscomp.Scope.Var;
 import com.google.javascript.jscomp.deps.SortedDependencies.CircularDependencyException;
 import com.google.javascript.jscomp.deps.SortedDependencies.MissingProvideException;
 import com.google.javascript.jscomp.mozilla.rhino.ErrorReporter;
@@ -173,8 +170,6 @@
 
   private final PrintStream outStream;
 
-  private GlobalVarReferenceMap globalRefMap = null;
-
   /**
    * Creates a Compiler that reports errors and warnings to its logger.
    */
@@ -258,20 +253,15 @@
           options.checkGlobalThisLevel);
     }
 
-    if (options.getLanguageIn() == LanguageMode.ECMASCRIPT5_STRICT) {
-      options.setWarningLevel(
-          DiagnosticGroups.ES5_STRICT,
-          CheckLevel.ERROR);
-    }
-
     // Initialize the warnings guard.
     List<WarningsGuard> guards = Lists.newArrayList();
     guards.add(
         new SuppressDocWarningsGuard(
             getDiagnosticGroups().getRegisteredGroups()));
-    guards.add(options.getWarningsGuard());
-
-    ComposeWarningsGuard composedGuards = new ComposeWarningsGuard(guards);
+    WarningsGuard warningsGuard = options.getWarningsGuard();
+    if (warningsGuard != null) {
+      guards.add(options.getWarningsGuard());
+    }
 
     // All passes must run the variable check. This synthesizes
     // variables later so that the compiler doesn't crash. It also
@@ -279,12 +269,13 @@
     // about missing variable declarations, we shut that specific
     // error off.
     if (!options.checkSymbols &&
-        !composedGuards.enables(DiagnosticGroups.CHECK_VARIABLES)) {
-      composedGuards.addGuard(new DiagnosticGroupWarningsGuard(
+        (warningsGuard == null || !warningsGuard.disables(
+            DiagnosticGroups.CHECK_VARIABLES))) {
+      guards.add(new DiagnosticGroupWarningsGuard(
           DiagnosticGroups.CHECK_VARIABLES, CheckLevel.OFF));
     }
 
-    this.warningsGuard = composedGuards;
+    this.warningsGuard = new ComposeWarningsGuard(guards);
   }
 
   /**
@@ -363,7 +354,6 @@
     // Create the source map if necessary.
     if (options.sourceMapOutputPath != null) {
       sourceMap = options.sourceMapFormat.getInstance();
-      sourceMap.setPrefixMappings(options.sourceMapLocationMappings);
     }
   }
 
@@ -765,7 +755,6 @@
     endPass();
   }
 
-  @Override
   void process(CompilerPass p) {
     p.process(externsRoot, jsRoot);
   }
@@ -909,7 +898,9 @@
     return errorManager.getWarnings();
   }
 
-  @Override
+  /**
+   * Returns the root node of the AST, which includes both externs and source.
+   */
   public Node getRoot() {
     return externAndJsRoot;
   }
@@ -963,25 +954,6 @@
     return inputsByName.get(name);
   }
 
-  /**
-   * Removes an input file from AST.
-   * @param name The name of the file to be removed.
-   */
-  protected void removeExternInput(String name) {
-    CompilerInput input = getInput(name);
-    if (input == null) {
-      return;
-    }
-    Preconditions.checkState(input.isExtern(), "Not an extern input: "
-        + input.getName());
-    inputsByName.remove(name);
-    externs.remove(input);
-    Node root = input.getAstRoot(this);
-    if (root != null) {
-      root.detachFromParent();
-    }
-  }
-
   @Override
   public CompilerInput newExternInput(String name) {
     if (inputsByName.containsKey(name)) {
@@ -991,17 +963,16 @@
     CompilerInput input = new CompilerInput(ast, name, true);
     inputsByName.put(name, input);
     externsRoot.addChildToFront(ast.getAstRoot(this));
-    externs.add(0, input);
     return input;
   }
 
   /** Add a source input dynamically. Intended for incremental compilation. */
   void addIncrementalSourceAst(JsAst ast) {
-    String intputName = ast.getSourceFile().getName();
+    String sourceName = ast.getSourceFile().getName();
     Preconditions.checkState(
-        getInput(intputName) == null,
-        "Duplicate input of name " + intputName);
-    inputsByName.put(intputName, new CompilerInput(ast));
+        getInput(sourceName) == null,
+        "Duplicate input of name " + sourceName);
+    inputsByName.put(sourceName, new CompilerInput(ast));
   }
 
   /**
@@ -1014,11 +985,11 @@
    * @return Whether the new AST was attached successfully.
    */
   boolean replaceIncrementalSourceAst(JsAst ast) {
-    String inputName = ast.getSourceFile().getName();
+    String sourceName = ast.getSourceFile().getName();
     CompilerInput oldInput =
         Preconditions.checkNotNull(
-            getInput(inputName),
-            "No input to replace: " + inputName);
+            getInput(sourceName),
+            "No input to replace: " + sourceName);
     Node newRoot = ast.getAstRoot(this);
     if (newRoot == null) {
       return false;
@@ -1032,7 +1003,7 @@
     }
 
     CompilerInput newInput = new CompilerInput(ast);
-    inputsByName.put(inputName, newInput);
+    inputsByName.put(sourceName, newInput);
 
     JSModule module = oldInput.getModule();
     if (module != null) {
@@ -1056,35 +1027,10 @@
   }
 
   @Override
-  MemoizedScopeCreator getTypedScopeCreator() {
+  ScopeCreator getTypedScopeCreator() {
     return getPassConfig().getTypedScopeCreator();
   }
 
-  @SuppressWarnings("unchecked")
-  DefaultPassConfig ensureDefaultPassConfig() {
-    PassConfig passes = getPassConfig().getBasePassConfig();
-    Preconditions.checkState(passes instanceof DefaultPassConfig,
-        "PassConfigs must eventually delegate to the DefaultPassConfig");
-    return (DefaultPassConfig) passes;
-  }
-
-  SymbolTable buildKnownSymbolTable() {
-    SymbolTable symbolTable = new SymbolTable();
-
-    MemoizedScopeCreator typedScopeCreator = getTypedScopeCreator();
-    if (typedScopeCreator != null) {
-      symbolTable.addSymbolsFrom(typedScopeCreator);
-    }
-
-    GlobalNamespace globalNamespace =
-        ensureDefaultPassConfig().getGlobalNamespace();
-    if (globalNamespace != null) {
-      symbolTable.addSymbolsFrom(globalNamespace);
-    }
-
-    return symbolTable;
-  }
-
   @Override
   public Scope getTopScope() {
     return getPassConfig().getTopScope();
@@ -1140,12 +1086,6 @@
     jsRoot = new Node(Token.BLOCK);
     jsRoot.setIsSyntheticBlock(true);
 
-    externsRoot = new Node(Token.BLOCK);
-    externsRoot.setIsSyntheticBlock(true);
-
-    externAndJsRoot = new Node(Token.BLOCK, externsRoot, jsRoot);
-    externAndJsRoot.setIsSyntheticBlock(true);
-
     if (options.tracer.isOn()) {
       tracker = new PerformanceTracker(jsRoot,
           options.tracer == TracerMode.ALL);
@@ -1156,6 +1096,8 @@
 
     try {
       // Parse externs sources.
+      externsRoot = new Node(Token.BLOCK);
+      externsRoot.setIsSyntheticBlock(true);
       for (CompilerInput input : externs) {
         Node n = input.getAstRoot(this);
         if (hasErrors()) {
@@ -1196,6 +1138,9 @@
       boolean staleInputs = false;
       for (CompilerInput input : inputs) {
         Node n = input.getAstRoot(this);
+        if (hasErrors()) {
+          return null;
+        }
 
         // Inputs can have a null AST during initial parse.
         if (n == null) {
@@ -1255,9 +1200,9 @@
         jsRoot.addChildToBack(n);
       }
 
-      if (hasErrors()) {
-        return null;
-      }
+      externAndJsRoot = new Node(Token.BLOCK, externsRoot, jsRoot);
+      externAndJsRoot.setIsSyntheticBlock(true);
+
       return externAndJsRoot;
     } finally {
       stopTracer(tracer, "parseInputs");
@@ -1439,7 +1384,7 @@
 
           String delimiter = options.inputDelimiter;
 
-          String sourceName = root.getSourceFileName();
+          String sourceName = (String)root.getProp(Node.SOURCENAME_PROP);
           Preconditions.checkState(sourceName != null);
           Preconditions.checkState(!sourceName.isEmpty());
 
@@ -1463,9 +1408,7 @@
               cb.getLineIndex(), cb.getColumnIndex());
         }
 
-        // if LanguageMode is ECMASCRIPT5_STRICT, only print 'use strict'
-        // for the first input file
-        String code = toSource(root, sourceMap, inputSeqNum == 0);
+        String code = toSource(root, sourceMap);
         if (!code.isEmpty()) {
           cb.append(code);
 
@@ -1494,19 +1437,19 @@
   @Override
   String toSource(Node n) {
     initCompilerOptionsIfTesting();
-    return toSource(n, null, true);
+    return toSource(n, null);
   }
 
   /**
    * Generates JavaScript source code for an AST.
    */
-  private String toSource(Node n, SourceMap sourceMap, boolean firstOutput) {
+  private String toSource(Node n, SourceMap sourceMap) {
     CodePrinter.Builder builder = new CodePrinter.Builder(n);
     builder.setPrettyPrint(options.prettyPrint);
     builder.setLineBreak(options.lineBreak);
     builder.setSourceMap(sourceMap);
     builder.setSourceMapDetailLevel(options.sourceMapDetailLevel);
-    builder.setTagAsStrict(firstOutput &&
+    builder.setTagAsStrict(
         options.getLanguageOut() == LanguageMode.ECMASCRIPT5_STRICT);
     builder.setLineLengthThreshold(options.lineLengthThreshold);
 
@@ -1709,12 +1652,7 @@
 
   @Override
   public boolean acceptEcmaScript5() {
-    switch (options.getLanguageIn()) {
-      case ECMASCRIPT5:
-      case ECMASCRIPT5_STRICT:
-        return true;
-    }
-    return false;
+    return options.getLanguageIn() == LanguageMode.ECMASCRIPT5;
   }
 
   public LanguageMode languageMode() {
@@ -1747,8 +1685,7 @@
       parserConfig = ParserRunner.createConfig(
         isIdeMode(),
         mode,
-        acceptConstKeyword(),
-        options.extraAnnotationNames);
+        acceptConstKeyword());
     }
     return parserConfig;
   }
@@ -1789,7 +1726,12 @@
   @Override
   public CheckLevel getErrorLevel(JSError error) {
     Preconditions.checkNotNull(options);
-    return warningsGuard.level(error);
+    WarningsGuard guards = options.getWarningsGuard();
+    if (guards == null) {
+      return error.level;
+    } else {
+      return guards.level(error);
+    }
   }
 
   /**
@@ -1951,13 +1893,6 @@
   }
 
   /**
-   * Gets the externs in the order in which they are being processed.
-   */
-  List<CompilerInput> getExternsInOrder() {
-    return Collections.<CompilerInput>unmodifiableList(externs);
-  }
-
-  /**
    * Stores the internal compiler state just before optimization is performed.
    * This can be saved and restored in order to efficiently optimize multiple
    * different output targets without having to perform checking multiple times.
@@ -2035,22 +1970,4 @@
   void setHasRegExpGlobalReferences(boolean references) {
     hasRegExpGlobalReferences = references;
   }
-
-  @Override
-  void updateGlobalVarReferences(Map<Var, ReferenceCollection> refMapPatch,
-      Node collectionRoot) {
-    Preconditions.checkState(collectionRoot.getType() == Token.SCRIPT
-        || collectionRoot.getType() == Token.BLOCK);
-    if (globalRefMap == null) {
-      globalRefMap = new GlobalVarReferenceMap(getInputsInOrder(),
-          getExternsInOrder());
-    }
-    globalRefMap.updateGlobalVarReferences(refMapPatch, collectionRoot);
-  }
-
-  @Override
-  ReferenceMap getGlobalVarReferences() {
-    return globalRefMap;
-  }
-
 }
diff -r -u buggy/test/com/google/javascript/jscomp/CommandLineRunnerTest.java inducing/test/com/google/javascript/jscomp/CommandLineRunnerTest.java
--- buggy/test/com/google/javascript/jscomp/CommandLineRunnerTest.java	2023-06-12 11:34:23.062158608 +0800
+++ inducing/test/com/google/javascript/jscomp/CommandLineRunnerTest.java	2023-06-12 11:34:23.270162410 +0800
@@ -18,12 +18,10 @@
 
 import com.google.common.base.Function;
 import com.google.common.base.Joiner;
-import com.google.common.base.Preconditions;
 import com.google.common.base.Supplier;
 import com.google.common.base.Suppliers;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
-import com.google.javascript.jscomp.CompilerOptions.LanguageMode;
 import com.google.javascript.rhino.Node;
 
 import junit.framework.TestCase;
@@ -116,30 +114,6 @@
     super.tearDown();
   }
 
-  public void testWarningGuardOrdering1() {
-    args.add("--jscomp_error=globalThis");
-    args.add("--jscomp_off=globalThis");
-    testSame("function f() { this.a = 3; }");
-  }
-
-  public void testWarningGuardOrdering2() {
-    args.add("--jscomp_off=globalThis");
-    args.add("--jscomp_error=globalThis");
-    test("function f() { this.a = 3; }", CheckGlobalThis.GLOBAL_THIS);
-  }
-
-  public void testWarningGuardOrdering3() {
-    args.add("--jscomp_warning=globalThis");
-    args.add("--jscomp_off=globalThis");
-    testSame("function f() { this.a = 3; }");
-  }
-
-  public void testWarningGuardOrdering4() {
-    args.add("--jscomp_off=globalThis");
-    args.add("--jscomp_warning=globalThis");
-    test("function f() { this.a = 3; }", CheckGlobalThis.GLOBAL_THIS);
-  }
-
   public void testCheckGlobalThisOffByDefault() {
     testSame("function f() { this.a = 3; }");
   }
@@ -159,22 +133,6 @@
          "function f(a) { return a; } f();");
   }
 
-  public void testReflectedMethods() {
-    args.add("--compilation_level=ADVANCED_OPTIMIZATIONS");
-    test(
-        "/** @constructor */" +
-        "function Foo() {}" +
-        "Foo.prototype.handle = function(x, y) { alert(y); };" +
-        "var x = goog.reflect.object(Foo, {handle: 1});" +
-        "for (var i in x) { x[i].call(x); }" +
-        "window['Foo'] = Foo;",
-        "function a() {}" +
-        "a.prototype.a = function(e, d) { alert(d); };" +
-        "var b = goog.c.b(a, {a: 1}),c;" +
-        "for (c in b) { b[c].call(b); }" +
-        "window.Foo = a;");
-  }
-
   public void testTypeCheckingOnWithVerbose() {
     args.add("--warning_level=VERBOSE");
     test("function f(x) { return x; } f();", TypeCheck.WRONG_ARGUMENT_COUNT);
@@ -224,12 +182,6 @@
     testSame("x = 3;");
   }
 
-  public void testCheckSymbolsOverrideForQuiet() {
-    args.add("--warning_level=QUIET");
-    args.add("--jscomp_error=undefinedVars");
-    test("x = 3;", VarCheck.UNDEFINED_VAR_ERROR);
-  }
-
   public void testCheckUndefinedProperties1() {
     args.add("--warning_level=VERBOSE");
     args.add("--jscomp_error=missingProperties");
@@ -298,22 +250,22 @@
 
   public void testProcessClosurePrimitives() {
     test("var goog = {}; goog.provide('goog.dom');",
-         "var goog = {dom:{}};");
+         "var goog = {}; goog.dom = {};");
     args.add("--process_closure_primitives=false");
     testSame("var goog = {}; goog.provide('goog.dom');");
   }
 
   public void testCssNameWiring() throws Exception {
-    test("var goog = {}; goog.getCssName = function() {};" +
-         "goog.setCssNameMapping = function() {};" +
-         "goog.setCssNameMapping({'goog': 'a', 'button': 'b'});" +
+    String prefix =
+        "var goog = {}; goog.getCssName = function() {};" +
+         "goog.setCssNameMapping = function() {};";
+    test(prefix + "goog.setCssNameMapping({'goog': 'a', 'button': 'b'});" +
          "var a = goog.getCssName('goog-button');" +
          "var b = goog.getCssName('css-button');" +
          "var c = goog.getCssName('goog-menu');" +
          "var d = goog.getCssName('css-menu');",
-         "var goog = { getCssName: function() {}," +
-         "             setCssNameMapping: function() {} }," +
-         "    a = 'a-b'," +
+         prefix +
+         "var a = 'a-b'," +
          "    b = 'css-b'," +
          "    c = 'a-menu'," +
          "    d = 'css-menu';");
@@ -336,7 +288,6 @@
 
   public void testIssue115() {
     args.add("--compilation_level=SIMPLE_OPTIMIZATIONS");
-    args.add("--jscomp_off=es5Strict");
     args.add("--warning_level=VERBOSE");
     test("function f() { " +
          "  var arguments = Array.prototype.slice.call(arguments, 0);" +
@@ -360,12 +311,6 @@
          "}");
   }
 
-  public void testIssue504() {
-    args.add("--compilation_level=ADVANCED_OPTIMIZATIONS");
-    test("void function() { alert('hi'); }();",
-         "alert('hi');", CheckSideEffects.USELESS_CODE_ERROR);
-  }
-
   public void testDebugFlag1() {
     args.add("--compilation_level=SIMPLE_OPTIMIZATIONS");
     args.add("--debug=false");
@@ -627,16 +572,7 @@
     args.add("--js_output_file");
     args.add("/path/to/out.js");
     testSame("var x = 3;");
-    assertEquals(SourceMap.Format.DEFAULT,
-        lastCompiler.getOptions().sourceMapFormat);
-  }
-
-  public void testSourceMapFormat2() {
-    args.add("--js_output_file");
-    args.add("/path/to/out.js");
-    args.add("--source_map_format=V3");
-    testSame("var x = 3;");
-    assertEquals(SourceMap.Format.V3,
+    assertEquals(SourceMap.Format.LEGACY,
         lastCompiler.getOptions().sourceMapFormat);
   }
 
@@ -654,8 +590,8 @@
           "var x = 3;", "var y = 5;", "var z = 7;", "var a = 9;"});
 
     StringBuilder builder = new StringBuilder();
-    lastCommandLineRunner.printModuleGraphManifestOrBundleTo(
-        lastCompiler.getModuleGraph(), builder, true);
+    lastCommandLineRunner.printModuleGraphManifestTo(
+        lastCompiler.getModuleGraph(), builder);
     assertEquals(
         "{m0}\n" +
         "i0\n" +
@@ -677,8 +613,8 @@
           "var x = 3;", "var y = 5;", "var z = 7;", "var a = 9;"});
 
     StringBuilder builder = new StringBuilder();
-    lastCommandLineRunner.printModuleGraphManifestOrBundleTo(
-        lastCompiler.getModuleGraph(), builder, true);
+    lastCommandLineRunner.printModuleGraphManifestTo(
+        lastCompiler.getModuleGraph(), builder);
     assertEquals(
         "{m0}\n" +
         "i0\n" +
@@ -777,48 +713,6 @@
        CheckAccessControls.DEPRECATED_NAME);
   }
 
-  public void testTwoParseErrors() {
-    // If parse errors are reported in different files, make
-    // sure all of them are reported.
-    Compiler compiler = compile(new String[] {
-      "var a b;",
-      "var b c;"
-    });
-    assertEquals(2, compiler.getErrors().length);
-  }
-
-  public void testES3ByDefault() {
-    test("var x = f.function", RhinoErrorReporter.PARSE_ERROR);
-  }
-
-  public void testES5() {
-    args.add("--language_in=ECMASCRIPT5");
-    test("var x = f.function", "var x = f.function");
-    test("var let", "var let");
-  }
-
-  public void testES5Strict() {
-    args.add("--language_in=ECMASCRIPT5_STRICT");
-    test("var x = f.function", "'use strict';var x = f.function");
-    test("var let", RhinoErrorReporter.PARSE_ERROR);
-  }
-
-  public void testES5StrictUseStrict() {
-    args.add("--language_in=ECMASCRIPT5_STRICT");
-    Compiler compiler = compile(new String[] {"var x = f.function"});
-    String outputSource = compiler.toSource();
-    assertEquals("'use strict'", outputSource.substring(0, 12));
-  }
-
-  public void testES5StrictUseStrictMultipleInputs() {
-    args.add("--language_in=ECMASCRIPT5_STRICT");
-    Compiler compiler = compile(new String[] {"var x = f.function",
-        "var y = f.function", "var z = f.function"});
-    String outputSource = compiler.toSource();
-    assertEquals("'use strict'", outputSource.substring(0, 12));
-    assertEquals(outputSource.substring(13).indexOf("'use strict'"), -1);
-  }
-
   /* Helper functions */
 
   private void testSame(String original) {
@@ -880,10 +774,6 @@
     test(new String[] { original }, warning);
   }
 
-  private void test(String original, String expected, DiagnosticType warning) {
-    test(new String[] { original }, new String[] { expected }, warning);
-  }
-
   /**
    * Asserts that when compiling, there is an error or warning.
    */
@@ -980,13 +870,8 @@
     for (int i = 0; i < original.length; i++) {
       inputs.add(JSSourceFile.fromCode("input" + i, original[i]));
     }
-    CompilerOptions options = new CompilerOptions();
-    // ECMASCRIPT5 is the most forgiving.
-    options.setLanguageIn(LanguageMode.ECMASCRIPT5);
-    compiler.init(externs, inputs, options);
+    compiler.init(externs, inputs, new CompilerOptions());
     Node all = compiler.parseInputs();
-    Preconditions.checkState(compiler.getErrorCount() == 0);
-    Preconditions.checkNotNull(all);
     Node n = all.getLastChild();
     return n;
   }
