diff -u -r data/changesInfo//Mockito_22/properties/modified_classes/inducing/src/org/mockito/internal/matchers/Equality.java data/changesInfo//Mockito_22/cleaned/fixing/src/org/mockito/internal/matchers/Equality.java
--- data/changesInfo//Mockito_22/properties/modified_classes/inducing/src/org/mockito/internal/matchers/Equality.java	2023-07-24 01:12:09.343269272 -0700
+++ data/changesInfo//Mockito_22/cleaned/fixing/src/org/mockito/internal/matchers/Equality.java	2023-07-24 02:15:18.919959224 -0700
@@ -6,6 +6,8 @@
 public class Equality {
 
     public static boolean areEqual(Object o1, Object o2) {
+        if (o1 == o2)
+            return true;
         if (o1 == null || o2 == null) {
             return o1 == null && o2 == null;
         } else if (isArray(o1)) {
Only in data/changesInfo//Mockito_22/properties/modified_classes/inducing/src/org/mockito/internal/matchers: Equals.java
Only in data/changesInfo//Mockito_22/properties/modified_classes/inducing/: test
