diff -u -r data/changesInfo//Time_2/properties/modified_classes/inducing/src/main/java/org/joda/time/field/UnsupportedDurationField.java data/changesInfo//Time_2/cleaned/fixing/src/main/java/org/joda/time/field/UnsupportedDurationField.java
--- data/changesInfo//Time_2/properties/modified_classes/inducing/src/main/java/org/joda/time/field/UnsupportedDurationField.java	2023-06-24 02:53:25.593520483 -0700
+++ data/changesInfo//Time_2/cleaned/fixing/src/main/java/org/joda/time/field/UnsupportedDurationField.java	2023-06-24 02:54:24.242032465 -0700
@@ -224,6 +224,9 @@
      * @return zero always
      */
     public int compareTo(DurationField durationField) {
+        if (durationField.isSupported()) {
+            return 1;
+        }
         return 0;
     }
 
Only in data/changesInfo//Time_2/properties/modified_classes/inducing/src: test
