diff -u -r data/changesInfo//Math_89/properties/modified_classes/inducing/src/java/org/apache/commons/math/stat/Frequency.java data/changesInfo//Math_89/cleaned/fixing/src/java/org/apache/commons/math/stat/Frequency.java
--- data/changesInfo//Math_89/properties/modified_classes/inducing/src/java/org/apache/commons/math/stat/Frequency.java	2023-12-27 22:48:22.302207699 -0800
+++ data/changesInfo//Math_89/cleaned/fixing/src/java/org/apache/commons/math/stat/Frequency.java	2023-06-24 01:37:12.227942507 -0700
@@ -107,7 +107,11 @@
      */
     @Deprecated
     public void addValue(Object v) {
-        addValue((Comparable<?>) v);
+        if (v instanceof Comparable<?>){
+            addValue((Comparable<?>) v);
+        } else {
+            throw new IllegalArgumentException("Object must implement Comparable");
+        }
     }
     
     /**
Only in data/changesInfo//Math_89/properties/modified_classes/inducing/src: test
