diff -u -r data/changesInfo//Lang_53/properties/modified_classes/inducing/src/java/org/apache/commons/lang/time/DateUtils.java data/changesInfo//Lang_53/cleaned/fixing/src/java/org/apache/commons/lang/time/DateUtils.java
--- data/changesInfo//Lang_53/properties/modified_classes/inducing/src/java/org/apache/commons/lang/time/DateUtils.java	2023-12-27 22:45:58.147773870 -0800
+++ data/changesInfo//Lang_53/cleaned/fixing/src/java/org/apache/commons/lang/time/DateUtils.java	2023-06-20 07:03:38.571838305 -0700
@@ -639,18 +639,18 @@
         int millisecs = val.get(Calendar.MILLISECOND);
         if (!round || millisecs < 500) {
             time = time - millisecs;
-            if (field == Calendar.SECOND) {
-                done = true;
-            }
+        }
+        if (field == Calendar.SECOND) {
+            done = true;
         }
 
         // truncate seconds
         int seconds = val.get(Calendar.SECOND);
         if (!done && (!round || seconds < 30)) {
             time = time - (seconds * 1000L);
-            if (field == Calendar.MINUTE) {
-                done = true;
-            }
+        }
+        if (field == Calendar.MINUTE) {
+            done = true;
         }
 
         // truncate minutes
Only in data/changesInfo//Lang_53/properties/modified_classes/inducing/src: test
