diff -u -r data/changesInfo//Time_1/properties/modified_classes/inducing/src/main/java/org/joda/time/Partial.java data/changesInfo//Time_1/cleaned/fixing/src/main/java/org/joda/time/Partial.java
--- data/changesInfo//Time_1/properties/modified_classes/inducing/src/main/java/org/joda/time/Partial.java	2023-12-27 22:48:33.766709856 -0800
+++ data/changesInfo//Time_1/cleaned/fixing/src/main/java/org/joda/time/Partial.java	2023-06-24 02:28:19.017718036 -0700
@@ -215,7 +215,7 @@
             DurationField loopUnitField = loopType.getDurationType().getField(iChronology);
             if (i > 0) {
                 int compare = lastUnitField.compareTo(loopUnitField);
-                if (compare < 0) {
+                if (compare < 0 || (compare != 0 && loopUnitField.isSupported() == false)) {
                     throw new IllegalArgumentException("Types array must be in order largest-smallest: " +
                             types[i - 1].getName() + " < " + loopType.getName());
                 } else if (compare == 0) {
Only in data/changesInfo//Time_1/properties/modified_classes/inducing/src: test
