diff -u -r data/changesInfo//Math_75/properties/modified_classes/inducing/src/java/org/apache/commons/math/stat/Frequency.java data/changesInfo//Math_75/cleaned/fixing/src/java/org/apache/commons/math/stat/Frequency.java
--- data/changesInfo//Math_75/properties/modified_classes/inducing/src/java/org/apache/commons/math/stat/Frequency.java	2023-12-27 22:48:03.369622838 -0800
+++ data/changesInfo//Math_75/cleaned/fixing/src/java/org/apache/commons/math/stat/Frequency.java	2023-06-24 01:03:11.848000000 -0700
@@ -299,7 +299,7 @@
      */
     @Deprecated
     public double getPct(Object v) {
-        return getCumPct((Comparable<?>) v);
+        return getPct((Comparable<?>) v);
     }
 
     /**
Only in data/changesInfo//Math_75/properties/modified_classes/inducing/src: main
Only in data/changesInfo//Math_75/properties/modified_classes/inducing/src: test
