diff -r -u buggy/src/com/google/javascript/jscomp/Compiler.java inducing/src/com/google/javascript/jscomp/Compiler.java
--- buggy/src/com/google/javascript/jscomp/Compiler.java	2023-06-12 11:34:24.426183543 +0800
+++ inducing/src/com/google/javascript/jscomp/Compiler.java	2023-06-12 11:34:24.486184640 +0800
@@ -265,20 +265,19 @@
             getDiagnosticGroups().getRegisteredGroups()));
     guards.add(options.getWarningsGuard());
 
-    ComposeWarningsGuard composedGuards = new ComposeWarningsGuard(guards);
-
     // All passes must run the variable check. This synthesizes
     // variables later so that the compiler doesn't crash. It also
     // checks the externs file for validity. If you don't want to warn
     // about missing variable declarations, we shut that specific
     // error off.
     if (!options.checkSymbols &&
-        !composedGuards.enables(DiagnosticGroups.CHECK_VARIABLES)) {
-      composedGuards.addGuard(new DiagnosticGroupWarningsGuard(
+        (warningsGuard == null || !warningsGuard.disables(
+            DiagnosticGroups.CHECK_VARIABLES))) {
+      guards.add(new DiagnosticGroupWarningsGuard(
           DiagnosticGroups.CHECK_VARIABLES, CheckLevel.OFF));
     }
 
-    this.warningsGuard = composedGuards;
+    this.warningsGuard = new ComposeWarningsGuard(guards);
   }
 
   /**
diff -r -u buggy/test/com/google/javascript/jscomp/CommandLineRunnerTest.java inducing/test/com/google/javascript/jscomp/CommandLineRunnerTest.java
--- buggy/test/com/google/javascript/jscomp/CommandLineRunnerTest.java	2023-06-12 11:34:24.426183543 +0800
+++ inducing/test/com/google/javascript/jscomp/CommandLineRunnerTest.java	2023-06-12 11:34:24.486184640 +0800
@@ -224,12 +224,6 @@
     testSame("x = 3;");
   }
 
-  public void testCheckSymbolsOverrideForQuiet() {
-    args.add("--warning_level=QUIET");
-    args.add("--jscomp_error=undefinedVars");
-    test("x = 3;", VarCheck.UNDEFINED_VAR_ERROR);
-  }
-
   public void testCheckUndefinedProperties1() {
     args.add("--warning_level=VERBOSE");
     args.add("--jscomp_error=missingProperties");
